src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/RemoteDiagnosticAnalyzerService.cs

```diff
@@ -96,6 +97,31 @@ public async ValueTask<SerializableDiagnosticAnalysisResults> CalculateDiagnosti
             }
         }
 
+        public async ValueTask<ImmutableArray<DiagnosticData>> GetSourceGeneratorDiagnosticsAsync(Checksum solutionChecksum, ProjectId projectId, CancellationToken cancellationToken)
+        {
+            return await RunWithSolutionAsync(
+                solutionChecksum,
+                async solution =>
+                {
+                    var project = solution.GetRequiredProject(projectId);
+                    var diagnostics = await project.GetSourceGeneratorDiagnosticsAsync(cancellationToken).ConfigureAwait(false);
+                    if (diagnostics.IsEmpty)
```

Comment:
don't need this fast path.  arraybuilders are already good at that.  also, if you commonly expect <4 diagnostics (which i would) TemporaryArray is a good choice.